src/content/ai/genui/input-events.md

```diff
@@ -0,0 +1,209 @@
+   it to the AI through the `ContentGenerator`,
+   which forwards it to the AI agent.
+
+## Defining events
+
+### Protocol level
+
+The A2UI protocol defines a `userAction` message used to report events.
+A `userAction` contains:
+
+* `name`: The name of the action
+  (defined by the AI when generating the component).
+* `surfaceId`: The ID of the UI surface where the event occurred.
+* `sourceComponentId`: The ID of the component that triggered the event.
+* `context`: A JSON object containing data relevant to the event.
+* `timestamp`: When the event occurred.
+
+### Dart implementation
+
+In [`packages/genui`][], user events are represented by the
+`UiEvent` extension type and its concrete implementation
+`UserActionEvent`.
+
+[`packages/genui`]: {{site.pub-pkg}}/genui
+
+The following structures are defined in
+`lib/src/model/ui_models.dart`:
+
+```dart
+/// A data object that represents a user interaction event in the UI.
+extension type UiEvent.fromMap(JsonMap _json) { ... }
+
+/// A UI event that represents a user action.
+extension type UserActionEvent.fromMap(JsonMap _json) implements UiEvent {
+  UserActionEvent({
+    String? surfaceId,
+    required String name,
+    required String sourceComponentId,
+    JsonMap? context,
+    // ...
+  }) : ...
+}
+```
+
+## Capturing events in widgets
+
+Widgets in GenUI are defined in a `Catalog`,
+which includes information about what events
+the widget can send to the AI.
+The AI can then send
+information about how to communicate those events back.
+When you implement a custom widget (or use the standard widgets),
+you use the `dispatchEvent` method in `CatalogItemContext`
+to dispatch events.
+
+### Example: Button implementation
+
+The following example shows how a `Button` widget typically captures
+a tap and dispatches an event. It retrieves the action definition
+(provided by the AI) from its properties,
+resolves any data bindings in the context, and sends the event.
+
+```dart
+// Inside a CatalogItem widgetBuilder
+widgetBuilder: (itemContext) {
+  // 1. Extract action data from the component properties
+  final buttonData = _ButtonData.fromMap(itemContext.data as JsonMap);
+  final JsonMap actionData = buttonData.action;
+  final actionName = actionData['name'] as String;
+
+  // 2. Extract context definition (which data to send back)
+  final List<Object?> contextDefinition =
+      (actionData['context'] as List<Object?>?) ?? <Object?>[];
+
+  return ElevatedButton(
+    onPressed: () {
+      // 3. Resolve the context values from the data model
+      final JsonMap resolvedContext = resolveContext(
+        itemContext.dataContext,
+        contextDefinition,
+      );
+
+      // 4. Dispatch the event
+      itemContext.dispatchEvent(
+        UserActionEvent(
+          name: actionName,
+          sourceComponentId: itemContext.id,
+          context: resolvedContext,
+        ),
+      );
+    },
+    child: ...
+  );
+},
+```
+
+## Event processing pipeline
```

Comment:
It seems you mentioned deleting this section earlier. Did that just refer to part of it or have you not pushed that change yet?